sys/kern/link_elf.c

```diff
@@ -2249,6 +2249,9 @@ elf_lookup_ifunc(linker_file_t lf, Elf_Size symidx, int deps __unused,
 	symp = ef->symtab + symidx;
 	if (ELF_ST_TYPE(symp->st_info) == STT_GNU_IFUNC) {
 		val = (caddr_t)ef->address + symp->st_value;
+#ifdef __CHERI_PURE_CAPABILITY__
+		val = make_capability(symp, val);
+#endif
 		*res = ((Elf_Addr (*)(void))val)();
```

Comment:
This, and res, should be using Elf_Addr. Do we not have any kernel IFUNCs on Morello? I guess this is only for relocations with symbols, but we wouldn't expect to see any of those in the kernel, everything should be non-preemptible and any references should be using IRELATIVE, which doesn't take a symbol.

Reply:
(That is, I do wonder in what instances this function is actually called dynamically upstream)

Reply:
Uh yes that's what I meant to write. We do have IFUNCs on Morello, but in-kernel references should always be non-preemptible and thus use IRELATIVE with no symbol, which means they won't take this path.

Reply:
But that's something that should be true on all architectures, not just Morello, so I'm confused as to why this function exists at all

Reply:
Oh, I did fix this to use uintptr_t in a later commit that changes the elf_lookup functions to return uintptr_t.

Reply:
Looks like the commit log introducing this function explains why there are non-IRELATIVE relocations.  I wonder if it is relevant for CHERI architectures though: 36716fe2e6eddac7ab1b9e9b1a268885a3985932

Reply:
I don't understand the justification there, unless it's just that kernel *modules* will have relocations against symbols (especially for relocatable objects on amd64)? Otherwise I still don't see how one could have relocations against symbols in the kernel itself.

Reply:
That I don't know, maybe @markjdb can comment?